### ISSUE TITLE
Seg fault on running using two processes each on one K80 GPU

### ISSUE THREAD

**Issue by such87** [sailorsb/caffe-parallel#8]
Hi,

I am trying to run caffe-parallel using 2 to 4 processes on a single node which has
8 K80 GPUs. The compilation is fine. I use gencode -arch sm_37 with CUDA 7.0.

I see a segmentation fault like follows :

[1] I1210 22:24:04.668552 39309 net.cpp:469] Collecting Learning Rate and Weight Decay.
[1] I1210 22:24:04.668563 39309 net.cpp:221] Network initialization done.
[1] I1210 22:24:04.668577 39309 net.cpp:222] Memory required for data: 31987608
[1] I1210 22:24:04.668613 39309 solver.cpp:245] Solver scaffolding done.
[1] I1210 22:24:04.668635 39309 solver.cpp:364] Solving CIFAR10_quick
[0] [node-0001:mpi_rank_0][error_sighandler] Caught error: Segmentation fault (signal 11)

=============================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 39308 RUNNING AT node-0001
=   EXIT CODE: 139
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
=============================================================================
 When I do a backtrace of the code dump I see the following :

#0  __pthread_mutex_lock (mutex=0x200000031) at pthread_mutex_lock.c:50
#1  0x00007ffecfb4a5c8 in boost::thread::native_handle() () from /usr/lib64/libboost_thread-mt.so.5
#2  0x0000000000527f91 in caffe::Thread::join() ()
#3  0x0000000000528039 in caffe::InternalThread::WaitForInternalThreadToExit() ()
#4  0x00000000004bc2c2 in caffe::BasePrefetchingDataLayer<float>::JoinPrefetchThread() ()
#5  0x0000000000512ca3 in caffe::BasePrefetchingDataLayer<float>::Forward_gpu_test(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> > const&, std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> >*) ()
#6  0x0000000000512f4f in caffe::BasePrefetchingDataLayer<float>::Forward_gpu_root(std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> > const&, std::vector<caffe::Blob<float>*, std::allocator<caffe::Blob<float>*> >*, int) ()
#7  0x0000000000469963 in caffe::Net<float>::ForwardPrefilledRoot(float*, int) ()
#8  0x00000000004874cc in caffe::Solver<float>::Solve(char const*) ()
#9  0x000000000041b484 in train() ()
#10 0x00000000004166e0 in main ()

Any ideas what may be going wrong ? I can provide any other detail regarding the execution
if you need. Please let me know.

Thanks
